test/jdk/java/awt/Focus/ComponentLostFocusTest.java

```diff
@@ -0,0 +1,138 @@
+import java.awt.FlowLayout;
+import java.awt.Frame;
+import java.awt.KeyboardFocusManager;
+import java.awt.Point;
+import java.awt.Robot;
+import java.awt.TextField;
+import java.awt.event.FocusAdapter;
+import java.awt.event.FocusEvent;
+import java.awt.event.InputEvent;
+import java.awt.event.WindowAdapter;
+import java.awt.event.WindowEvent;
+
+public class ComponentLostFocusTest {
+
+    Frame frame = new Frame("Frame");
+    TextField tf = new TextField("Text Field");
+    static Robot r;
+    Dialog dialog = null;
+    volatile boolean passed;
+
+    public ComponentLostFocusTest() {
+        try{
+            r = new Robot();
+        } catch(Exception e){
+            throw new RuntimeException(e);
+        }
+        r.setAutoDelay(100);
+
+        dialog = new Dialog(frame, "Dialog", true);
+
+        frame.addWindowFocusListener(new WindowAdapter() {
+            public void windowGainedFocus(WindowEvent e) {
+                System.out.println("Frame gained focus: "+e);
+            }
+        });
+
+        frame.setLayout (new FlowLayout ());
+        frame.add(tf);
+        frame.setSize(400,300);
+        frame.setVisible(true);
+        frame.setLocationRelativeTo(null);
+        frame.validate();
+    }
+
+    public void doTest() {
+        // Do requesting focus to the modal dialog in order to after that
+        // to do requesting focus to the frame
+        System.out.println("dialog.setVisible.... ");
+        new Thread(new Runnable() {
+            public void run() {
+                dialog.setVisible(true);
+            }
+        }).start();
+
+        // The bug is that this construction leads to the redundant xRequestFocus
+        // By the way, the requestFocusInWindow() works fine before the fix
+        System.out.println("requesting.... ");
+        frame.requestFocus();
+
+        r.delay(1000);
+
+        // Returning the focus to the initial frame will work correctly after the fix
+        System.out.println("disposing.... ");
+        dialog.dispose();
+
+        r.delay(1000);
+
+        // We want to track the GAIN_FOCUS from this time
+        tf.addFocusListener(new FocusAdapter() {
+            public void focusGained(FocusEvent e) {
+                System.out.println("TextField gained focus: "+e);
+                passed = true;
+            }
+        });
+
+        doRequestFocusToTextField();
+
+        System.out.println("Focused window: " + KeyboardFocusManager.getCurrentKeyboardFocusManager().getFocusedWindow());
+        System.out.println("Focus owner: " + KeyboardFocusManager.getCurrentKeyboardFocusManager().getFocusOwner());
+
+        if (!passed) {
+            throw new RuntimeException("TextField got no focus! Test failed.");
+        }
+    }
+
+    private void doRequestFocusToTextField(){
+        // do activation using press title
+        Point loc = frame.getLocationOnScreen();
+        r.mouseMove(loc.x + frame.getWidth()/2, loc.y + frame.getInsets().top/2);
```

Comment:
Should it be accessed on EDT?

Reply:
modified test to put relevant portion under EDT and also added dispose..